src/engine.c

```diff
@@ -869,6 +869,79 @@ ENGINE_ellipse(ENGINE* engine, int64_t x0, int64_t y0, int64_t x1, int64_t y1, u
   ENGINE_blitBuffer(engine, x0, y0);
 }
 
+internal void
+ENGINE_triangle(ENGINE* engine, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t c) {
+  ENGINE_line(engine, x0, y0, x1, y1, c, 1);
+  ENGINE_line(engine, x1, y1, x2, y2, c, 1);
+  ENGINE_line(engine, x2, y2, x0, y0, c, 1);
+}
+
+internal void
+ENGINE_trianglefillFlatBottom(ENGINE* engine, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t c) {
+  float invslope0 = (x1 - x0) / (y1 - y0);
+  float invslope1 = (x2 - x0) / (y2 - y0);
+  float curx0 = x0;
+  float curx1 = x0;
+
+  for (int scanlineY = y0; scanlineY <= y1; scanlineY++) {
+    ENGINE_line(engine, curx0, scanlineY, curx1, scanlineY, c, 1);
+    curx0 += invslope0;
+    curx1 += invslope1;
+  }
+}
+
+internal void
+ENGINE_trianglefillFlatTop(ENGINE* engine, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t c) {
+  float invslope0 = (x2 - x0) / (y2 - y0);
+  float invslope1 = (x2 - x1) / (y2 - y1);
+
+  float curx0 = x2;
+  float curx1 = x2;
+
+  for (int scanlineY = y2; scanlineY > y0; scanlineY--)
+  {
+    ENGINE_line(engine, curx0, scanlineY, curx1, scanlineY, c, 1);
+    curx0 -= invslope0;
+    curx1 -= invslope1;
+  }
+}
+
+internal void
+ENGINE_trianglefill(ENGINE* engine, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t c) {
+  if (y1 < y0) {
+    swap(&x1, &x0);
+    swap(&y1, &y0);
+  }
+  if (y2 < y0) {
+    swap(&x2, &x0);
+    swap(&y2, &y0);
+  }
+  if (y2 < y1) {
+    swap(&x2, &x1);
+    swap(&y2, &y1);
+  }
+
+  x0 += 0.5;
+  y0 += 0.5;
+  x1 += 0.5;
+  y1 += 0.5;
+  x2 += 0.5;
+  y2 += 0.5;
+
+  if (y1 == y2) {
+    ENGINE_trianglefillFlatBottom(engine, x0, y0, x1, y1, x2, y2, c);
+  } else if (y0 == y1) {
+    ENGINE_trianglefillFlatTop(engine, x0, y0, x1, y1, x2, y2, c);
+  } else {
+    float x3 = (x0 + ((y1 - y0) / (y2 - y0)) * (x2 - x0));
+    ENGINE_trianglefillFlatBottom(engine, x0, y0, x1, y1, x3, y1, c);
+    ENGINE_trianglefillFlatTop(engine, x1, y1, x3, y1, x2, y2, c);
+  }
+  ENGINE_print(engine, "p0", x0, y0, 0xff00ccff);
+  ENGINE_print(engine, "p1", x1, y1, 0xff00ccff);
+  ENGINE_print(engine, "p2", x2, y2, 0xff00ccff);
```

Comment:
Don't forget to remove these, as they are likely for debugging.

Reply:
Will do before marking this ready :+1: